### PR TITLE
Add docs for generate policies on existing resources

### LINKED ISSUE #536: [Enhancement] Add docs for generateExistingOnPolicyUpdate policies
### Description

KDP: https://github.com/kyverno/KDP/pull/10
Implementation: https://github.com/kyverno/kyverno/pull/3819, https://github.com/kyverno/kyverno/pull/3635

## Example Policies

1. This example generate data `NetworkPolicy` to both incoming and existing Namespace with `generateExistingOnPolicyUpdate: true`

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: "..."   
spec:
  background: true
  validationFailureAction: audit
  generateExistingOnPolicyUpdate: true   // trigger generate rule in existing resources
  rules:
  - name: default-deny
    match:
      resources:
        kinds:
        - Namespace
    generate:
      kind: NetworkPolicy
      name: default-deny
      namespace: "{{request.object.metadata.name}}"
      synchronize: true
      data:
        spec:
          # select all pods in the namespace
          podSelector: {}
          # deny all traffic
          policyTypes:
          - Ingress
          - Egress
```

2. This example generate clone `Secret` to both incoming and existing Namespace with `genExisting=true`

```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: "..."
spec:
  background: true
  validationFailureAction: audit
  generateExistingOnPolicyUpdate: true   // trigger generate rule in existing resources
  rules:
  - name: sync-image-pull-secret
    match:
      resources:
        kinds:
        - Namespace
    generate:
      apiVersion: v1
      kind: Secret
      name: secret-basic-auth
      namespace: "{{request.object.metadata.name}}"
      synchronize: true
      clone:
        namespace: default
        name: secret-basic-auth
```

### Slack discussion

_No response_

### PR DESCRIPTION
**Pull request by prateekpandey14**
Signed-off-by: prateekpandey14 <prateekpandey14@gmail.com>

## Related issue #

Closes #536

## Proposed Changes

Add docs for generate policies on existing resources

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/kyverno/website/blob/main/CONTRIBUTING.md).
- [x] I have inspected the website preview for accuracy.
- [x] I have signed off my issue.

### CODE REVIEW

File: content/en/docs/Writing policies/generate.md [hunk 1]
```diff
@@ -246,7 +246,130 @@ spec:
...
+      name: "{{request.object.metadata.name}}-default-pdb"
+      namespace: "{{request.object.metadata.namespace}}"
+      synchronize: true
+      data:
+        spec:
+          minAvailable: 1
+          selector:
+            matchLabels:
+              "{{request.object.metadata.labels}}"
+
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:347]: Remove the extra line.

File: content/en/docs/Writing policies/generate.md [hunk 2]
```diff
@@ -246,7 +246,130 @@ spec:
...
+```
+
+###
+To troubleshoot the policy application failure, you can inspect `UpdateRequest` Custom Resource  to get details. Successful `UpdateRequests` are automatically cleaned up by Kyverno.
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:351]: "Successful `UpdateRequests` are automatically cleaned up by Kyverno."

I'm not sure if this is the case for generate, can you please confirm?

File: content/en/docs/Writing policies/generate.md [hunk 3]
```diff
@@ -246,7 +246,130 @@ spec:
...
+...
+
+status:
+  message: 'poddisruptionbudgets.policy is forbidden: User "system:serviceaccount:kyverno:kyverno-service-account"
+            cannot create resource "poddisruptionbudgets" in API group "policy" in the namespace "test"'
+ state: Failed
+```
+
+
+## Generating resources into existing namespaces using Selector
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:372]: Do we want to mark this as deprecated as your change introduced a general solution to generate on existing resources?

File: content/en/docs/Writing policies/generate.md [hunk 4]
```diff
@@ -246,7 +246,130 @@ spec:
...
+      data:
+        spec:
+          minAvailable: 1
+          selector:
+            matchLabels:
+              "{{request.object.metadata.labels}}"
+
+```
+
+###
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:350]: Missing title?

File: content/en/docs/Writing policies/generate.md (hunk 2 above)
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:351]: No needs to be removed

File: content/en/docs/Writing policies/generate.md (hunk 4 above)
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:350]: Added

File: content/en/docs/Writing policies/generate.md [hunk 5]
```diff
@@ -246,7 +246,130 @@ spec:
           foo: bar
 ```
 
-## Generating resources into existing namespaces
+## Generate Existing resources
+
+With Kyverno 1.7.0+, Kyverno supports the generate existing resources. Unlike standard generate policies that are applied through the AdmissionReview process, generate existing policies are applied in the background which creates existing resources in the cluster. These generate existing policies, like traditional generate policies, are still triggered via the AdmissionReview process but apply to existing--and even different--resources. They may also optionally be configured to apply upon updates to the policy itself.
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md]: Right , i took some reference from the mutate existing PR , and missed to clean some of them.

File: content/en/docs/Writing policies/generate.md [hunk 6]
```diff
@@ -246,7 +246,111 @@ spec:
...
 
-## Generating resources into existing namespaces
+## Generate for Existing resources
+
+With Kyverno 1.7.0+, Kyverno supports the generate existing resources. Generate existing policies are applied in the background which creates target resource in existing matched resources in the cluster.
+They may also optionally be configured to apply upon updates to the policy itself.
+
+{{% alert title="Note" color="warning" %}}
+The proper permissions may need to be granted to Kyverno ServiceAccount. Kubernetes resources may require granting additional privileges to the ClusterRole responsible for the generate behavior. To enable Kyverno to generate these other types, edit the ClusterRole typically named kyverno:generate and add or update the rules to cover the resources and verbs needed.
+{{% /alert %}}
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:256]: This applies to all generate policies, not just "generate for existing". Do we want to add the note in this section? We also clarify this in the installation guide.

File: content/en/docs/Writing policies/generate.md [hunk 7]
```diff
@@ -246,7 +246,111 @@ spec:
           foo: bar
 ```
 
-## Generating resources into existing namespaces
+## Generate for Existing resources
+
+With Kyverno 1.7.0+, Kyverno supports the generate existing resources. Generate existing policies are applied in the background which creates target resource in existing matched resources in the cluster.
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:251]: ...supports the generate **for** existing resources...

File: content/en/docs/Writing policies/generate.md [hunk 8]
```diff
@@ -246,7 +246,111 @@ spec:
...
+
+{{% alert title="Note" color="warning" %}}
+The proper permissions may need to be granted to Kyverno ServiceAccount. Kubernetes resources may require granting additional privileges to the ClusterRole responsible for the generate behavior. To enable Kyverno to generate these other types, edit the ClusterRole typically named kyverno:generate and add or update the rules to cover the resources and verbs needed.
+{{% /alert %}}
+
+### Generate NetworkPolicy on Existing Namespaces
+
+By default, policy will not be applied on existing trigger resource when it is installed. This behavior can be configured via `generateExistingOnPolicyUpdate` attribute. If you set `generateExistingOnPolicyUpdate` to `true`, Kyverno will generate the target secret resource in existing namespaces on policy CREATE and UPDATE events.
+
+In this example policy, which matches the trigger resource kind `Namespace` and generate data `NetworkPolicy` on all the existing or newly created `Namespace`
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:262]: Missing period `.`.

File: content/en/docs/Writing policies/generate.md [hunk 9]
```diff
@@ -246,7 +246,111 @@ spec:
...
+## Generate for Existing resources
+
+With Kyverno 1.7.0+, Kyverno supports the generate existing resources. Generate existing policies are applied in the background which creates target resource in existing matched resources in the cluster.
+They may also optionally be configured to apply upon updates to the policy itself.
+
+{{% alert title="Note" color="warning" %}}
+The proper permissions may need to be granted to Kyverno ServiceAccount. Kubernetes resources may require granting additional privileges to the ClusterRole responsible for the generate behavior. To enable Kyverno to generate these other types, edit the ClusterRole typically named kyverno:generate and add or update the rules to cover the resources and verbs needed.
+{{% /alert %}}
+
+### Generate NetworkPolicy on Existing Namespaces
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:258]: Do we want to use the sub-title for example policies?

File: content/en/docs/Writing policies/generate.md [hunk 10]
```diff
@@ -246,7 +246,111 @@ spec:
...
+status:
+  message: 'poddisruptionbudgets.policy is forbidden: User "system:serviceaccount:kyverno:kyverno-service-account"
+            cannot create resource "poddisruptionbudgets" in API group "policy" in the namespace "test"'
+ state: Failed
+```
+
+
+## Generating resources into existing namespaces using Selector (Deprecated)
+
+With Kyverno 1.7.0+, match selector is deprecated and support to [generate existing resources](#generate-for-existing-resources) using a Policy has been added.
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:353]: Suggest changing to:

This feature has been deprecated in Kyverno 1.7.0+, refer to [this](#generate-for-existing-resources) section to generate for existing resources.

File: content/en/docs/Writing policies/generate.md [hunk 11]
```diff
@@ -246,7 +246,111 @@ spec:
...
+With Kyverno 1.7.0+, Kyverno supports the generate existing resources. Generate existing policies are applied in the background which creates target resource in existing matched resources in the cluster.
+They may also optionally be configured to apply upon updates to the policy itself.
+
+{{% alert title="Note" color="warning" %}}
+The proper permissions may need to be granted to Kyverno ServiceAccount. Kubernetes resources may require granting additional privileges to the ClusterRole responsible for the generate behavior. To enable Kyverno to generate these other types, edit the ClusterRole typically named kyverno:generate and add or update the rules to cover the resources and verbs needed.
+{{% /alert %}}
+
+### Generate NetworkPolicy on Existing Namespaces
+
+By default, policy will not be applied on existing trigger resource when it is installed. This behavior can be configured via `generateExistingOnPolicyUpdate` attribute. If you set `generateExistingOnPolicyUpdate` to `true`, Kyverno will generate the target secret resource in existing namespaces on policy CREATE and UPDATE events.
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:260]: Should we point out that this feature won't work if `generateExistingOnPolicyUpdate` is not set to `true`?

File: content/en/docs/Writing policies/generate.md (hunk 10 above)
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:353]: done

File: content/en/docs/Writing policies/generate.md (hunk 11 above)
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:260]: added the `Only if you set ...` part , should be good enough.

File: content/en/docs/Writing policies/generate.md (hunk 9 above)
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:258]: With sub-title its looks more clear and easy to understand, WDYT ?

Comment (realshuting) [reply content/en/docs/Writing policies/generate.md:254]: I'm fine with this.

Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:254]: Also we have 2 different policies , so having with sub-titles helps

File: content/en/docs/Writing policies/generate.md [hunk 12]
```diff
@@ -246,7 +246,107 @@ spec:
...
 
-## Generating resources into existing namespaces
+## Generate for Existing resources
+
+With Kyverno 1.7.0+, Kyverno supports the generate for existing resources. Generate existing policies are applied in the background which creates target resource in existing matched resources in the cluster.
+They may also optionally be configured to apply upon updates to the policy itself.
+
+### Generate NetworkPolicy on Existing Namespaces
+
+By default, policy will not be applied on existing trigger resource when it is installed. This behavior can be configured via `generateExistingOnPolicyUpdate` attribute. Only if you set `generateExistingOnPolicyUpdate` to `true`, Kyverno will generate the target secret resource in existing namespaces on policy CREATE and UPDATE events.
```
Comment (realshuting) [inline content/en/docs/Writing policies/generate.md:256]: Suggest removing specific Kind in this statement:

Kyverno will generate the target ~~secret~~ resources for existing triggers on ...

**Review by realshuting — COMMENTED** [review]
Looks good to me, needs small tweaks.

@chipzoller to provide final review.

File: content/en/docs/Writing policies/generate.md [hunk 13]
```diff
@@ -26,16 +26,16 @@ Kubernetes has many default resource types even before considering CustomResourc
 When generating a custom resource, it is necessary to set the apiVersion (ex., `spec.generate.apiVersion` and kind (ex., `spec.generate.kind`).
 {{% /alert %}}
 
-Kyverno will create an intermediate object called a `GenerateRequest` which is used to queue work items for the final resource generation. To get the details and status of a generated resource, check the details of the `GenerateRequest`. The following will give the list of `GenerateRequests`.
+Kyverno will create an intermediate object called a `UpdateRequest` which is used to queue work items for the final resource generation. To get the details and status of a generated resource, check the details of the `UpdateRequest`. The following will give the list of `UpdateRequests`.
 
 ```sh
 kubectl get generaterequests -A
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:32]: This command has not been updated accordingly.

File: content/en/docs/Writing policies/generate.md [hunk 14]
```diff
@@ -246,7 +246,107 @@ spec:
           foo: bar
 ```
 
-## Generating resources into existing namespaces
+## Generate for Existing resources
+
+With Kyverno 1.7.0+, Kyverno supports the generate for existing resources. Generate existing policies are applied in the background which creates target resource in existing matched resources in the cluster.
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:251]: Second sentence is slightly confusing. Suggest the following: "Generate existing policies are applied in the background which creates target resources based on the match statement within the policy."

File: content/en/docs/Writing policies/generate.md [hunk 15]
```diff
@@ -246,7 +246,107 @@ spec:
...
+They may also optionally be configured to apply upon updates to the policy itself.
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:252]: No need for a new paragraph here. Put this sentence with the previous paragraph.

File: content/en/docs/Writing policies/generate.md [hunk 16]
```diff
@@ -246,7 +246,107 @@ spec:
...
+
+### Generate NetworkPolicy on Existing Namespaces
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:254]: Should be "in" and not "on"

File: content/en/docs/Writing policies/generate.md [hunk 17]
```diff
@@ -246,7 +246,107 @@ spec:
...
+          labels:
+            created-by: kyverno
+        spec:
+          podSelector: {}
+          policyTypes:
+          - Ingress
+          - Egress
+```
+
+### Generate PodDisruptionBudget for Existing Deployment
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:290]: Should be "Deployments" (plural)

File: content/en/docs/Writing policies/generate.md [hunk 18]
```diff
@@ -246,7 +246,107 @@ spec:
...
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: create-default-pdb
+spec:
+  generateExistingOnPolicyUpdate: true
+  rules:
+  - name: create-default-pdb
+    match:
+      resources:
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:304]: Needs `any` | `all`

File: content/en/docs/Writing policies/generate.md [hunk 19]
```diff
@@ -246,7 +246,107 @@ spec:
...
+      synchronize: true
+      data:
+        spec:
+          minAvailable: 1
+          selector:
+            matchLabels:
+              "{{request.object.metadata.labels}}"
+```
+
+### Troubleshoot
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:325]: "Troubleshooting"

Comment (chipzoller) [reply content/en/docs/Writing policies/generate.md:325]: Space between heading and first sentence.

File: content/en/docs/Writing policies/generate.md [hunk 20]
```diff
@@ -246,7 +246,107 @@ spec:
...
+Name:         ur-7gtbx
+Namespace:    kyverno
+...
+
+status:
+  message: 'poddisruptionbudgets.policy is forbidden: User "system:serviceaccount:kyverno:kyverno-service-account"
+            cannot create resource "poddisruptionbudgets" in API group "policy" in the namespace "test"'
+ state: Failed
+```
+
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:345]: Remove excess new lines. Should be one new line only.

**Comment by prateekpandey14** [comment]
> Inline comments.

fixed

File: content/en/docs/Writing policies/generate.md [hunk 21]
```diff
@@ -246,7 +246,106 @@ spec:
...
+      data:
+        spec:
+          minAvailable: 1
+          selector:
+            matchLabels:
+              "{{request.object.metadata.labels}}"
+```
+
+### Troubleshooting
+To troubleshoot policy application failures, inspect the `UpdateRequest` Custom Resource to get details.
```
Comment (chipzoller) [inline content/en/docs/Writing policies/generate.md:326]: New line between heading and paragraph.

**Review by chipzoller — CHANGES_REQUESTED** [review]
Last couple things.

File: content/en/docs/Writing policies/generate.md [hunk 22]
```diff
@@ -246,7 +246,106 @@ spec:
           foo: bar
 ```
 
-## Generating resources into existing namespaces
```
[the remark this answers is not in the document]
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:249]: Sure , done

File: content/en/docs/Writing policies/generate.md (hunk 21 above)
Comment (prateekpandey14) [reply content/en/docs/Writing policies/generate.md:327]: done

**Pull request merged**